File: backend/replay_test.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	larkim "github.com/larksuite/oapi-sdk-go/v3/service/im/v1"
)

type original struct {
	WhenMs int64  `json:"whenMs"`
	Title  string `json:"title"`
	Text   string `json:"text"`
}

// corpusSource is a Lark that answers from a pulled corpus: what chats a title names, which
// p2p chat a Sender has, and what each chat said.
type corpusSource struct {
	originals []original
	titles    map[string][]string
	dms       map[string]string
	messages  map[string][]Candidate
}

func (c *corpusSource) GroupChatIDs(_ context.Context, title string) ([]string, error) {
	return c.titles[title], nil
}
func (c *corpusSource) DMChatID(_ context.Context, p Preview, whenMs int64) (string, error) {
	return c.dms[p.Sender], nil
}
func (c *corpusSource) Messages(_ context.Context, chatID string, whenMs int64) ([]Candidate, error) {
	return c.messages[chatID], nil
}

func readJSON(t *testing.T, dir, name string, into any) {
	data, err := os.ReadFile(filepath.Join(dir, name))
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(data, into); err != nil {
		t.Fatalf("%s: %v", name, err)
	}
}

func jsonLines(t *testing.T, dir, name string) [][]byte {
	data, err := os.ReadFile(filepath.Join(dir, name))
	if err != nil {
		t.Fatal(err)
	}
	var out [][]byte
	for _, line := range bytes.Split(data, []byte("\n")) {
		if len(bytes.TrimSpace(line)) > 0 {
			out = append(out, line)
		}
	}
	return out
}

// loadCorpus reads what `tools/larklish-helper replay fetch` wrote. The messages are the raw
// API items, decoded into the SDK's struct and turned into Candidates by the same candidateOf
// the live Source uses — so the replay measures the shipped rules.
func loadCorpus(t *testing.T, dir string) *corpusSource {
	src := &corpusSource{titles: map[string][]string{}, dms: map[string]string{}, messages: map[string][]Candidate{}}
	for _, line := range jsonLines(t, dir, "originals.jsonl") {
		var o original
		if err := json.Unmarshal(line, &o); err != nil {
			t.Fatal(err)
		}
		src.originals = append(src.originals, o)
	}
	readJSON(t, dir, "titles.json", &src.titles)
	readJSON(t, dir, "dms.json", &src.dms)
	for _, line := range jsonLines(t, dir, "messages.jsonl") {
		var m larkim.Message
		if err := json.Unmarshal(line, &m); err != nil {
			t.Fatal(err)
		}
		chat := str(m.ChatId)
		src.messages[chat] = append(src.messages[chat], candidateOf(&m))
	}
	return src
}

// TestReplayCorpus replays a day of real Originals through the Lookup the Backend ships, over
// a corpus pulled by `tools/larklish-helper replay fetch`. The corpus is real colleagues'
// messages, so it is never committed; without it this test skips. It writes one line per
// Original to outcomes-go.tsv (the Layer 7 gate diffed it against the Kotlin replay's).
func TestReplayCorpus(t *testing.T) {
	dir := "../replay-corpus"
	if _, err := os.Stat(filepath.Join(dir, "originals.jsonl")); err != nil {
		t.Skip("no corpus — run `tools/larklish-helper replay fetch`")
	}
	src := loadCorpus(t, dir)
	f := quiet(src)
	var lines []string
	counts := map[string]int{}
	for i, o := range src.originals {
		pick := f.FullTextOf(t.Context(), o.Title, ParsePreview(o.Text), o.WhenMs)
		if pick.Found != nil {
			lines = append(lines, fmt.Sprintf("%d\tfound\t%s\t%d", i, pick.ChatID, pick.Found.CreateTime))
			counts["found"]++
		} else {
			lines = append(lines, fmt.Sprintf("%d\t%s\t\t", i, pick.Reason))
			counts[pick.Reason]++
		}
	}
	if err := os.WriteFile(filepath.Join(dir, "outcomes-go.tsv"), []byte(strings.Join(lines, "\n")+"\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	reasons := make([]string, 0, len(counts))
	for r := range counts {
		reasons = append(reasons, r)
	}
	sort.Slice(reasons, func(i, j int) bool { return counts[reasons[i]] > counts[reasons[j]] })
	t.Logf("replay: %d of %d Originals resolved", counts["found"], len(src.originals))
	for _, r := range reasons {
		t.Logf("  %4d  %s", counts[r], r)
	}
	// Experiment 11 measured 119 of 210 for these rules; a real regression drops far below half.
	if counts["found"]*2 < len(src.originals) {
		t.Errorf("only %d of %d resolved", counts["found"], len(src.originals))
	}
}

File: app/src/test/kotlin/com/vegerot/larklish/ReplayTest.kt
package com.vegerot.larklish

import java.io.File
import org.junit.Assert.assertTrue
import org.junit.Assume.assumeTrue
import org.junit.Test

// The separators `tools/larklish-helper replay fetch` writes: items, paragraphs, fields of an
// element.
private const val SEP_ITEM = '\u0001'
private const val SEP_PARA = '\u0002'
private const val SEP_FIELD = '\u0003'

/** Undo the escaping the corpus applies to every leaf string. */
private fun String.unescaped(): String {
    val out = StringBuilder(length)
    var i = 0
    while (i < length) {
        val c = this[i]
        if (c == '\\' && i + 1 < length) {
            i++
            out.append(
                when (this[i]) {
                    'n' -> '\n'
                    't' -> '\t'
                    else -> this[i]
                }
            )
        } else {
            out.append(c)
        }
        i++
    }
    return out.toString()
}

private fun String.items() = split(SEP_ITEM).filter { it.isNotEmpty() }

private data class Original(val whenMs: Long, val title: String, val text: String)

/**
 * Replays a day of real Originals against the rules the app ships, over a corpus pulled by
 * `tools/larklish-helper replay fetch`. Experiments 10 and 11 scored the same thing in Python,
 * which measured a *copy* of the rules; this measures `pickMessage`, `postText`, `unwrapHtml`,
 * `resolveMentions` and `Preview.parse` themselves.
 *
 * The corpus is real colleagues' messages, so it is never committed. Without it this test skips.
 * Pull one, then read the breakdown:
 *
 *     tools/larklish-helper replay fetch
 *     ./gradlew testDebugUnitTest --tests '*ReplayTest*' -i
 */
class ReplayTest {
    private val corpus: File? =
        sequenceOf(
                System.getProperty("larklish.corpus"),
                "../replay-corpus",
                "replay-corpus",
            )
            .filterNotNull()
            .map(::File)
            .firstOrNull { File(it, "originals.tsv").exists() }

    private fun rows(name: String) =
        File(corpus, name).readLines().filter { it.isNotBlank() }.map { it.split("\t") }

    private fun candidatesByChat(): Map<String, List<Candidate>> =
        rows("messages.tsv").groupBy({ it[0] }) { r ->
            val kind = r[1]
            val deleted = r[3] == "1"
            val mentions =
                r[4].items().associate {
                    val f = it.split(SEP_FIELD)
                    f[0].unescaped() to f[1].unescaped()
                }
            val payload = r.getOrElse(5) { "" }
            val text =
                when {
                    deleted -> ""
                    kind == "text" -> unwrapHtml(payload.unescaped())
                    kind == "post" ->
                        payload.split(SEP_PARA).let { parts ->
                            postText(
                                parts[0].unescaped(),
                                parts.drop(1).map { para ->
                                    para.items().map {
                                        val f = it.split(SEP_FIELD)
                                        PostElement(
                                            f[0].unescaped(),
                                            f[1].unescaped(),
                                            f[2].unescaped(),
                                            f[3].unescaped(),
                                        )
                                    }
                                },
                            )
                        }
                    else -> ""
                }
            Candidate("", kind, r[2].toLong(), deleted, resolveMentions(text, mentions))
        }

    @Test
    fun theShippedRulesFindTheMessageBehindMostOriginals() {
        assumeTrue("no corpus — run `tools/larklish-helper replay fetch`", corpus != null)
        val messages = candidatesByChat()
        val titles =
            rows("titles.tsv").associate { it[0].unescaped() to it.getOrElse(1) { "" }.items() }
        val dms = rows("dms.tsv").associate { "dm:" + it[0].unescaped() to it[1] }
        val originals =
            rows("originals.tsv").map {
                Original(it[0].toLong(), it[1].unescaped(), it[2].unescaped())
            }

        // The policy MessageFetcher.fullTextOf runs, over a corpus instead of the network.
        val cache = HashMap<String, String>()
        val recent = ArrayDeque<String>()
        fun pickIn(chatId: String, stem: String, whenMs: Long): Pick =
            pickMessage(messages[chatId].orEmpty(), stem, whenMs).also {
                if (it is Pick.Found) {
                    recent.remove(chatId)
                    recent.addFirst(chatId)
                }
            }

        val rows = originals.map { o ->
            val preview = Preview.parse(o.text)
            val isDm = o.title == "Lark" || o.title == preview.sender
            val key = if (isDm) "dm:" + preview.sender else o.title
            val candidates =
                cache[key]?.let { listOf(it) }
                    ?: if (isDm) listOfNotNull(dms[key]) else titles[o.title].orEmpty()
            var outcome: Pick = Pick.Skipped("no-chat")
            var hit: Pick? = null
            var hitChat = ""
            candidates.forEachIndexed { i, chatId ->
                if (hit == null) {
                    val pick = pickIn(chatId, preview.stem, o.whenMs)
                    if (i == 0) outcome = pick
                    if (pick is Pick.Found) {
                        cache[key] = chatId
                        hit = pick
                        hitChat = chatId
                    }
                }
            }
            if (hit == null && candidates.size == 1) cache[key] = candidates[0]
            if (hit == null) {
                for (chatId in recent.take(RECENT_IN_REPLAY)) {
                    val pick = pickIn(chatId, preview.stem, o.whenMs)
                    if (pick is Pick.Found) {
                        hit = pick
                        hitChat = chatId
                        break
                    }
                }
            }
            when (val p = hit ?: outcome) {
                is Pick.Found -> "found\t$hitChat\t${p.candidate.createTime}"
                is Pick.Skipped -> "${p.reason}\t\t"
            }
        }
        // One line per Original — index, outcome, chat, create_time — for the diff against the
        // Go replay (Layer 7). Message ids are not in the corpus, so the chat and time name the pick.
        File(corpus, "outcomes-kotlin.tsv").writeText(rows.mapIndexed { i, r -> "$i\t$r\n" }.joinToString(""))
        val outcomes = rows.map { it.substringBefore("\t") }

        val found = outcomes.count { it == "found" }
        println("replay: $found of ${originals.size} Originals resolved")
        outcomes
            .groupingBy { it }
            .eachCount()
            .toList()
            .sortedByDescending { it.second }
            .forEach { (reason, n) -> println("  ${n.toString().padStart(4)}  $reason") }
        // Experiment 11 measured 119 of 210 for these rules. The floor allows for a corpus
        // pulled over a different span; a real regression drops it far below this.
        assertTrue("only $found of ${originals.size} resolved", found * 2 >= originals.size)
    }
}

/** Mirrors `MessageFetcher.RECENT_TRIES`, which is private to the main source set. */
private const val RECENT_IN_REPLAY = 2
